### PR TITLE
dkg: check disk writes

### PR DESCRIPTION
**Pull request by dB2510**
Check disk writes by creating random sample files and then removing sample ones if successful.

category: bug
ticket: #584

### CODE REVIEW

File: dkg/disk.go [hunk 1]
```diff
@@ -112,10 +113,32 @@ func writeDepositData(aggSigs map[core.PubKey]*bls_sig.Signature, withdrawalAddr
...
+	err = os.WriteFile(depositPath, bytes, 0o444) // read-only
 	if err != nil {
 		return errors.Wrap(err, "write deposit data")
 	}
 
 	return nil
 }
+
+// fileExists returns error if keystores, cluster-lock and deposit-data already exists in given dataDir.
+func fileExists(dataDir string, vals int) error {
```
[the remark this answers is not in the document]
Comment (corverroos) [reply dkg/disk.go:125]: as per ticket:
```
The command should check for clashes (and all other disk write issues) at the start and fail with a proper message.
```

File: dkg/disk.go [hunk 2]
```diff
@@ -112,10 +116,74 @@ func writeDepositData(aggSigs map[core.PubKey]*bls_sig.Signature, withdrawalAddr
...
+		return errors.Wrap(err, "remove sample deposit-data.json")
+	}
+
+	for i := 0; i < def.NumValidators; i++ {
+		if err := os.Remove(path.Join(dataDir, fmt.Sprintf("keystore-%d.json", i))); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("remove sample keystore-%d.json", i))
+		}
+
+		if err := os.Remove(path.Join(dataDir, fmt.Sprintf("keystore-%d.txt", i))); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("remove sample keystore-%d.txt", i))
```
Comment (corverroos) [inline dkg/disk.go:180]: avoid sprintf in errors, rather use `z.Str("file", filename)`

File: dkg/disk.go [hunk 3]
```diff
@@ -112,10 +116,74 @@ func writeDepositData(aggSigs map[core.PubKey]*bls_sig.Signature, withdrawalAddr
...
+
+		sig, err := testutil.RandomBLSSignature()
+		if err != nil {
+			return err
+		}
+
+		sigs[pk] = sig
+	}
+
+	if err := writeDepositData(sigs, "0x0000000000000000000000000000000000000000", "prater", dataDir); err != nil {
```
Comment (corverroos) [inline dkg/disk.go:156]: no const for this withdrawal addr avialble?

File: dkg/disk.go [hunk 4]
```diff
@@ -112,10 +116,74 @@ func writeDepositData(aggSigs map[core.PubKey]*bls_sig.Signature, withdrawalAddr
...
+		return err
```
Comment (corverroos) [inline dkg/disk.go:157]: Are these error message human friendly? Will people understand what the problem is?

File: dkg/disk.go [hunk 5]
```diff
@@ -112,10 +116,74 @@ func writeDepositData(aggSigs map[core.PubKey]*bls_sig.Signature, withdrawalAddr
...
 	}
 
 	return nil
 }
+
+// checkWrites writes sample files to check disk writes and removes sample files after verification.
+func checkWrites(dataDir string, def cluster.Definition) error {
+	var shares []share
+	sigs := make(map[core.PubKey]*bls_sig.Signature)
+	for i := 0; i < def.NumValidators; i++ {
```
Comment (corverroos) [inline dkg/disk.go:131]: another option which is simpler is:
```
files := getSampleFiles(def.NumValidators)
for f in files {
  err := os.WriteFile("sample", f, 0x444)
  if err ! nil { "cannot write file", z.Str("file",f) }
  
  err := os.Remove(f)
  if err ! nil { "cannot delete file", z.Str("file",f) }
} 
```

**Review by corverroos — APPROVED** [review]

File: dkg/disk.go (hunk 3 above)
Comment (dB2510) [reply dkg/disk.go:156]: can be replaced with `testutil.RandomETHAddress()`

**Pull request merged**